core/vpn/src/network.rs

```diff
@@ -558,68 +439,143 @@ impl Handler<Disconnect> for Vpn {
-    type Result = <DataSent as Message>::Result;
+/// Handle ingress event from stack
+impl Handler<Ingress> for Vpn {
+    type Result = ActorResponse<Self, Result<()>>;
 
-    fn handle(&mut self, _: DataSent, ctx: &mut Self::Context) -> Self::Result {
-        self.poll(ctx.address());
-        Ok(())
+    fn handle(&mut self, msg: Ingress, ctx: &mut Self::Context) -> Self::Result {
+        match msg.event {
+            IngressEvent::InboundConnection { desc } => {
+                log::debug!(
+                    "[vpn] ingress: connection to {:?} ({}) from {:?}",
+                    desc.local,
+                    desc.protocol,
+                    desc.remote
+                );
+                ActorResponse::reply(Ok(()))
+            }
+            IngressEvent::Disconnected { desc } => {
+                log::debug!(
+                    "[vpn] ingress: disconnect {:?} ({}) by {:?}",
+                    desc.local,
+                    desc.protocol,
+                    desc.remote,
+                );
+                ctx.address()
+                    .do_send(Disconnect::new(desc, DisconnectReason::SocketClosed));
+
+                ActorResponse::reply(Ok(()))
+            }
+            IngressEvent::Packet { payload, desc, .. } => {
+                if let Some(mut connection) = self.connections.get(&desc).cloned() {
+                    log::debug!("[vpn] ingress proxy: send to {:?}", desc.local);
+
+                    let fut = async move { connection.ingress_tx.send(payload).await }
+                        .into_actor(self)
+                        .map(move |res, _, ctx| {
+                            res.map_err(|e| {
+                                ctx.address()
+                                    .do_send(Disconnect::new(desc, DisconnectReason::SinkClosed));
+
+                                Error::Other(e.to_string())
+                            })
+                        });
+                    ActorResponse::r#async(fut)
+                } else {
+                    log::debug!("[vpn] ingress proxy: no connection to {:?}", desc);
+                    ActorResponse::reply(Ok(()))
+                }
+            }
+        }
+    }
+}
+
+/// Handle egress event from stack
+impl Handler<Egress> for Vpn {
+    type Result = ActorResponse<Self, Result<()>>;
+
+    fn handle(&mut self, msg: Egress, _: &mut Self::Context) -> Self::Result {
+        let frame = msg.event.payload.into_vec();
+
+        log::debug!("[vpn] egress -> runtime packet {} B", frame.len());
+
+        match self.vpn.endpoint(msg.event.remote) {
+            Some(endpoint) => {
+                let fut = endpoint
+                    .udp
+                    .push_raw_as(&self.node_id, frame)
+                    .map(|r| match r {
+                        Ok(_) => Ok(()),
+                        Err(e) => Err(Error::Other(e.to_string())),
+                    });
+
+                ActorResponse::r#async(fut.into_actor(self))
```

Comment:
You could try comparing performance of this solution vs using `ctx.spawn(fut.into_actor(self));`